hott/cubical/squareover.lean

```diff
@@ -0,0 +1,288 @@
+    exact eq_of_fn_eq_fn (pathover_equiv_tr_eq _ _ _) H'
+  end
+
+  /- A version of eq_pathover where the type of the equality also varies -/
+  @[hott] lemma eq_pathover_dep {f g : Πa, B a} {p : a = a'} {q : f a = g a}
+    {r : f a' = g a'} (s : squareover B hrfl (pathover_idp_of_eq q) (pathover_idp_of_eq r)
+                                             (apd f p) (apd g p)) : q =[p; λx, f x = g x] r :=
+  begin
+    induction p, apply pathover_idp_of_eq, apply eq_of_vdeg_square, exact square_of_squareover_ids s
+  end
+
+  /- charcaterization of pathovers in pathovers -/
+  -- in this version the fibration (B) of the pathover does not depend on the variable (a)
+   @[hott] lemma pathover_pathover {a' a₂' : A'} {p : a' = a₂'} {f g : A' → A}
+    {b : Πa, B (f a)} {b₂ : Πa, B (g a)} {q : Π(a' : A'), f a' = g a'}
+    (r : pathover B (b a') (q a') (b₂ a'))
+    (r₂ : pathover B (b a₂') (q a₂') (b₂ a₂'))
+    (s : squareover B (natural_square q p) r r₂
+                      (pathover_ap B f (apd b p)) (pathover_ap B g (apd b₂ p)))
+    : pathover (λa, pathover B (b a) (q a) (b₂ a)) r p r₂ :=
+  begin
+    induction p, apply pathover_idp_of_eq, apply eq_of_vdeg_squareover, exact s
+  end
+
+  @[hott] def squareover_change_path_left {p₀₁' : a₀₀ = a₀₂} (r : p₀₁' = p₀₁)
+    {q₀₁ : b₀₀ =[p₀₁'] b₀₂} (t : squareover B (r ⬝ph s₁₁) q₁₀ q₁₂ q₀₁ q₂₁)
+  : squareover B s₁₁ q₁₀ q₁₂ (change_path r q₀₁) q₂₁ :=
+  by induction r; exact t
+
+  @[hott] def squareover_change_path_right {p₂₁' : a₂₀ = a₂₂} (r : p₂₁' = p₂₁)
+    {q₂₁ : b₂₀ =[p₂₁'] b₂₂} (t : squareover B (s₁₁ ⬝hp r⁻¹) q₁₀ q₁₂ q₀₁ q₂₁)
+  : squareover B s₁₁ q₁₀ q₁₂ q₀₁ (change_path r q₂₁) :=
+  by induction r; exact t
+
+  @[hott] def squareover_change_path_right' {p₂₁' : a₂₀ = a₂₂} (r : p₂₁ = p₂₁')
+    {q₂₁ : b₂₀ =[p₂₁'] b₂₂} (t : squareover B (s₁₁ ⬝hp r) q₁₀ q₁₂ q₀₁ q₂₁)
+  : squareover B s₁₁ q₁₀ q₁₂ q₀₁ (change_path r⁻¹ q₂₁) :=
+  by induction r; exact t
+
+  /- You can construct a square in a sigma-type by giving a squareover -/
+  @[hott] def square_dpair_eq_dpair {a₀₀ a₂₀ a₀₂ a₂₂ : A}
+    {p₁₀ : a₀₀ = a₂₀} {p₀₁ : a₀₀ = a₀₂} {p₂₁ : a₂₀ = a₂₂} {p₁₂ : a₀₂ = a₂₂}
+    (s₁₁ : square p₁₀ p₁₂ p₀₁ p₂₁) {b₀₀ : B a₀₀} {b₂₀ : B a₂₀} {b₀₂ : B a₀₂} {b₂₂ : B a₂₂}
+    {q₁₀ : b₀₀ =[p₁₀] b₂₀} {q₀₁ : b₀₀ =[p₀₁] b₀₂} {q₂₁ : b₂₀ =[p₂₁] b₂₂} {q₁₂ : b₀₂ =[p₁₂] b₂₂}
+    (t₁₁ : squareover B s₁₁ q₁₀ q₁₂ q₀₁ q₂₁) :
+    square (sigma.dpair_eq_dpair p₁₀ q₁₀) (dpair_eq_dpair p₁₂ q₁₂)
+           (dpair_eq_dpair p₀₁ q₀₁) (dpair_eq_dpair p₂₁ q₂₁) :=
+  by induction t₁₁; constructor
+
+  @[hott] lemma sigma_square {v₀₀ v₂₀ v₀₂ v₂₂ : Σa, B a}
+    {p₁₀ : v₀₀ = v₂₀} {p₀₁ : v₀₀ = v₀₂} {p₂₁ : v₂₀ = v₂₂} {p₁₂ : v₀₂ = v₂₂}
+    (s₁₁ : square p₁₀..1 p₁₂..1 p₀₁..1 p₂₁..1)
+    (t₁₁ : squareover B s₁₁ p₁₀..2 p₁₂..2 p₀₁..2 p₂₁..2) : square p₁₀ p₁₂ p₀₁ p₂₁ :=
+  begin
+    induction v₀₀, induction v₂₀, induction v₀₂, induction v₂₂,
+    rwr [(sigma_eq_eta p₁₀)⁻¹, (sigma_eq_eta p₀₁)⁻¹, (sigma_eq_eta p₁₂)⁻¹, (sigma_eq_eta p₂₁)⁻¹],
```

Comment:
The hott `rwr` tactic has exactly the same syntax as the standard `rw` tactic, in particular you can use the left-arrow to rewrite right-to-left:
```lean
rwr [← sigma_eq_eta p₁₀, ...]
```